Exceptions.md

```diff
@@ -0,0 +1,199 @@
+While a client is debugging their app, it would be useful for them to get a stack trace in case of unhandled exceptions which goes all the way through the stack, and is not interrupted by the Dafny layer.
+
+
+## Implementation
+
+The type `Result<T>` is hardcoded into Dafny, the same way as `seq`, `map`, `int`, etc.
+
+Given a `MethodCall(args)` which returns `Result<T>`, let's define
+
+```
+var v :- MethodCall(args);
+... rest of the code using v as if it had type T ...
+```
+
+to be syntactic sugar for
+
+```
+var res_v: Result<T> := MethodCall(args);
+if res_v.Failure? {
+	return Failure(res_v.error);
+}
+var v := res_v.value;
+... rest of the code using v as if it had type T ...
+```
+
+as far as verification and Boogie is concerned.
+
+We will not really implement it as syntactic sugar, though, because we want a separate AST node for this so that the compiler can regognize it easily.
+
+The compiler will then compile all methods which return `Result<T>` into C# methods with signatures of the form
+
+```
+public T MethodCall(A1 a1, A2 a2, ...) throws Exception
+```
+
+(Note that C# method signatures do not have `throws` clauses, but we just add them here for clarity, even though that's not valid C# code).
+
+In particular, the same signature translation also applies for methods annotated with `{:extern}`, so for instance, given the C# method `System.IO.Stream.Read` with signature 
+
+```
+public int Read (byte[] buffer, int offset, int count)
+```
+
+one could write a Dafny stub for it as
+
+```
+method {:extern} Read(buffer: array<byte>, offset: int32, count: int32) returns (nRead: Result<int32>)
+```
+
+which would enable us to call exception-throwing C# methods from Dafny without an extra unverified layer turning exceptions into `Result<int32>`.
+
+In the same way, methods implemented in Dafny which return `Result<T>` would be compiled to exception-throwing methods and thus be directly usable by C# client programs.
+
+Moreover,
+
+```
+return Failure(someErrorMsg);
+```
+
+is compiled into a `throw` statement:
+
+```
+throw new Exception(dafnyStringToNativeString(someErrorMsg));
+```
+
+Compilation of
+
+```
+return Success(someValue);
+```
+
+simply drops the wrapping:
+
+```
+return someValue;
+```
+
+And the new Dafny construct
+
+```
+var value1: T1 :- operation1(args1);
+```
+
+is simply compiled into
+
+```
+var value1 = operation1(args1);
+```
+
+and the handling of the failure case is done by C#'s exception mechanism.
+
+
+## FAQ
+
+*Q:* Are we also adding special syntax/support for this inside expressions?
+*A:* Not at the moment
+
+*Q:* What about target languages which do not have exceptions, such as Go?
+*A:* Go uses a pattern of returning an additional return value to report success/failure, and the Dafny-to-Go compiler would use this pattern too. In general, the idea is to compile `Result<T>` to the idiomatic way of dealing with failure in the target language, be it exceptions, extra return values, `Option`, `Maybe`, ...
+
```

Comment:
Yep. In Go, the signature of the target method would have 2 out-parameters, one for the success value (intended to be used only if the second value is non-`nil`) and an outcome indication (which is the error, or is `nil` for success).

Go sometimes uses a boolean instead of an `error`. It would be good if the Dafny-to-Go compiler treated this case, too.